sopel/modules/dice.py

```diff
@@ -99,89 +99,153 @@ def get_compressed_string(self):
+    """Custom exception type for failing to drop lowest N dice."""
+    def __init__(self, dropped: int, total: int):
+        super().__init__(dropped, total)
+
+    @property
+    def dropped(self) -> int:
+        return self.args[0]
+
+    @property
+    def total(self) -> int:
+        return self.args[1]
+
+
+def _get_error_message(exc: DiceError) -> str:
+    if isinstance(exc, InvalidDiceFacesError):
+        return "I don't have any dice with {} sides.".format(exc.faces)
+    if isinstance(exc, NegativeDiceCountError):
+        return "I can't roll {} dice.".format(exc.count)
+    if isinstance(exc, TooManyDiceError):
+        return "I only have {}/{} dice.".format(exc.available, exc.requested)
+    if isinstance(exc, UnableToDropDiceError):
+        return "I can't drop the lowest {} of {} dice.".format(
+            exc.dropped, exc.total)
+
+    return "Unknown error rolling dice: %r" % exc
+
+
+def _roll_dice(dice_match: re.Match[str]) -> DicePouch:
+    dice_num = int(dice_match.group('dice_num') or 1)
+    dice_type = int(dice_match.group('dice_type'))
 
     # Dice can't have zero or a negative number of sides.
     if dice_type <= 0:
-        bot.reply("I don't have any dice with %d sides. =(" % dice_type)
-        return None  # Signal there was a problem
+        raise InvalidDiceFacesError(dice_type)
 
     # Can't roll a negative number of dice.
     if dice_num < 0:
-        bot.reply("I'd rather not roll a negative amount of dice. =(")
-        return None  # Signal there was a problem
+        raise NegativeDiceCountError(dice_num)
 
     # Upper limit for dice should be at most a million. Creating a dict with
     # more than a million elements already takes a noticeable amount of time
     # on a fast computer and ~55kB of memory.
-    if dice_num > 1000:
-        bot.reply('I only have 1000 dice. =(')
-        return None  # Signal there was a problem
+    if dice_num > MAX_DICE:
+        raise TooManyDiceError(dice_num, MAX_DICE)
 
-    dice = DicePouch(dice_num, dice_type, 0)
+    dice = DicePouch(dice_num, dice_type)
 
-    if result.group('drop_lowest'):
-        drop = int(result.group('drop_lowest'))
+    if dice_match.group('drop_lowest'):
+        drop = int(dice_match.group('drop_lowest'))
         if drop >= 0:
             dice.drop_lowest(drop)
         else:
-            bot.reply("I can't drop the lowest %d dice. =(" % drop)
+            raise UnableToDropDiceError(drop, dice_num)
 
     return dice
 
 
 @plugin.command('roll', 'dice', 'd')
 @plugin.priority("medium")
+@plugin.example(".roll", "No dice to roll.")
+@plugin.example(".roll 2d6+4^2&",
+                "I don't know how to process that. "
+                "Are the dice as well as the algorithms correct?")
+@plugin.example(".roll 65(2)", "I couldn't find any valid dice expressions.")
+@plugin.example(".roll 1d0", "I don't have any dice with 0 sides.")
+@plugin.example(".roll -1d6", "I can't roll -1 dice.")
```

Comment:
Both are good but can you do `.roll 1d-2`?

Reply:
The check is `<= 0` and the `1d0` example hits that. But it's definitely not hard to add an explicit test case for negative faces if you think it's important.

Reply:
Well it's digital dices, who know what they can do or represent?

Also, I'm pretty sure it'll raise a different error than the 1d0 case!

Reply:
> Also, I'm pretty sure it'll raise a different error than the 1d0 case!

It doesn't change the error (except for the number). This would be the new example, which I just tested locally:

```py
@plugin.example(".roll 1d-2", "I don't have any dice with -2 sides.")  # verified would pass locally
```

Did you mean `.roll -1d-2`, maybe? For that, `_roll_dice()` checks for `InvalidDiceFacesError` before `NegativeDiceCountError`, so the output is the same as it would be for `.roll 1d-2`.

```py
@plugin.example(".roll -1d-2", "I don't have any dice with -2 sides.")  # verified would pass locally
```

Reply:
Oh! I thought the regex didn't catch the negative sign in front of the dice. Well, apparently it does! It is a different case than 0 anyway, because if we do `<= 0` at the moment, it doesn't mean we'll always do that in the future. And negative number, 0, and positive number are 3 different cases.

Reply:
Added a `2d-2` case just for you. 🙂